app/modules/uart.c

```diff
@@ -7,91 +7,65 @@
 #include "c_types.h"
 #include "c_string.h"
 #include "rom.h"
+#include "driver/input.h"
 
 static int uart_receive_rf = LUA_NOREF;
-bool run_input = true;
-bool uart_on_data_cb(const char *buf, size_t len){
-  if(!buf || len==0)
-    return false;
-  if(uart_receive_rf == LUA_NOREF)
-    return false;
+
+void uart_on_data_cb(const char *buf, size_t len){
   lua_State *L = lua_getstate();
-  if(!L)
-    return false;
   lua_rawgeti(L, LUA_REGISTRYINDEX, uart_receive_rf);
   lua_pushlstring(L, buf, len);
-  lua_call(L, 1, 0);
-  return !run_input;
+  luaN_call(L, 1, 0, 0);
 }
 
-uint16_t need_len = 0;
-int16_t end_char = -1;
 // Lua: uart.on("method", [number/char], function, [run_input])
 static int l_uart_on( lua_State* L )
 {
-  size_t sl, el;
-  int32_t run = 1;
-  uint8_t stack = 1;
-  const char *method = luaL_checklstring( L, stack, &sl );
-  stack++;
-  if (method == NULL)
-    return luaL_error( L, "wrong arg type" );
-
-  if( lua_type( L, stack ) == LUA_TNUMBER )
+  size_t el;
+  int stack = 2, data_len = -1;
+  char end_char = 0;
+  const char *method = lua_tostring( L, 1);
+  bool run_input = true;
+  luaL_argcheck(L, method && !c_strcmp(method, "data"), 1, "method not supported");
+
+  if (lua_type( L, stack ) == LUA_TNUMBER)
   {
-    need_len = ( uint16_t )luaL_checkinteger( L, stack );
+    data_len = luaL_checkinteger( L, stack );
+    luaL_argcheck(L, data_len >= 0 && data_len <= 255, stack, "wrong arg range");
     stack++;
-    end_char = -1;
-    if( need_len > 255 ){
-      need_len = 255;
-      return luaL_error( L, "wrong arg range" );
-    }
   }
-  else if(lua_isstring(L, stack))
+  else if (lua_isstring(L, stack))
   {
     const char *end = luaL_checklstring( L, stack, &el );
+    data_len = 0;
+    end_char = (int16_t) end[0];
     stack++;
-    if(el!=1){
+    if(el!=1) {
       return luaL_error( L, "wrong arg range" );
     }
-    end_char = (int16_t)end[0];
-    need_len = 0;
   }
 
-  // luaL_checkanyfunction(L, stack);
-  if (lua_type(L, stack) == LUA_TFUNCTION || lua_type(L, stack) == LUA_TLIGHTFUNCTION){
-    if ( lua_isnumber(L, stack+1) ){
-      run = lua_tointeger(L, stack+1);
+  if (lua_isfunction(L, stack) || lua_islightfunction(L, stack)) {
```

Comment:
I believe you added a nice `lua_isanyfunction(L, stack)` macro above.

Reply:
Yup, but I'll be doing all of these as a separate sweep.

Reply:
Noted. And thumbs up.